state-sync/storage-service/server/src/moderator.rs

```diff
@@ -0,0 +1,395 @@
+    max_invalid_requests: u64, // The max number of invalid requests before ignoring the peer
+    min_time_to_ignore_secs: u64, // The min time (secs) to ignore the peer (doubles each round)
+    time_service: TimeService, // The time service
+}
+
+impl UnhealthyPeerState {
+    pub fn new(
+        max_invalid_requests: u64,
+        min_time_to_ignore_secs: u64,
+        time_service: TimeService,
+    ) -> Self {
+        Self {
+            ignore_start_time: None,
+            invalid_request_count: 0,
+            max_invalid_requests,
+            min_time_to_ignore_secs,
+            time_service,
+        }
+    }
+
+    /// Increments the invalid request count for the peer and marks
+    /// the peer to be ignored if it has sent too many invalid requests.
+    /// Note: we only ignore peers on the public network.
+    pub fn increment_invalid_request_count(&mut self, peer_network_id: &PeerNetworkId) {
+        // Increment the invalid request count
+        self.invalid_request_count += 1;
+
+        // If the peer is a PFN and has sent too many invalid requests, start ignoring it
+        if self.ignore_start_time.is_none()
+            && peer_network_id.network_id().is_public_network()
+            && self.invalid_request_count >= self.max_invalid_requests
+        {
+            // TODO: at some point we'll want to terminate the connection entirely
+
+            // Start ignoring the peer
+            self.ignore_start_time = Some(self.time_service.now());
+
+            // Log the fact that we're now ignoring the peer
+            warn!(LogSchema::new(LogEntry::RequestModeratorIgnoredPeer)
+                .peer_network_id(peer_network_id)
+                .message("Ignoring peer due to too many invalid requests!"));
+        }
+    }
+
+    /// Returns true iff the peer should be ignored
+    pub fn is_ignored(&self) -> bool {
+        self.ignore_start_time.is_some()
+    }
+
+    /// Refreshes the peer's state (if it has been ignored for long enough).
+    /// Note: each time we unblock a peer, we double the min time to ignore the peer.
+    /// This provides an exponential backoff for peers that are sending too many invalid requests.
+    pub fn refresh_peer_state(&mut self, peer_network_id: &PeerNetworkId) {
+        if let Some(ignore_start_time) = self.ignore_start_time {
+            let ignored_duration = self.time_service.now().duration_since(ignore_start_time);
+            if ignored_duration >= Duration::from_secs(self.min_time_to_ignore_secs) {
+                // Reset the invalid request count
+                self.invalid_request_count = 0;
+
+                // Reset the ignore start time
+                self.ignore_start_time = None;
+
+                // Double the min time to ignore the peer
+                self.min_time_to_ignore_secs *= 2;
+
+                // Log the fact that we're no longer ignoring the peer
+                warn!(LogSchema::new(LogEntry::RequestModeratorIgnoredPeer)
```

Comment:
Yeah, for now I want to keep the "start ignoring" and "stop ignoring" logs at the WARN level, simply so that we can see them in mainnet (where INFO won't be visible). Once we're confident that this logic works and bad behaviors are dealt with correctly, we can tune them back down. 😄